File: Code/Avernakis Nodejs/Test-Nodejs/src/Ave/Ui/Ribbon/UiRibbonBackstage.ts
import { IControl } from "../UiControl";
import { AveLib } from "../../AveLib";
import { Window } from "../Control/UiWindow";
import { IconCache, Rect, DpiSize } from "../UiCommon";

export interface IRibbonBackstage extends IControl {
    new (window: Window, key?: string | number): IRibbonBackstage;

    SetText(s: string): RibbonBackstage;
    GetText(): string;

    SetCount(nCount: number): RibbonBackstage;
    GetCount(): number;
    Add(): RibbonBackstage;
    Select(nIndex: number): RibbonBackstage;
    GetSelection(): number;
    GetHighlight(): number;

    SetText(nIndex: number, s: string): RibbonBackstage;
    GetText(nIndex: number): string;
    SetIcon(nIndex: number, nIcon: IconCache): RibbonBackstage;
    GetIcon(nIndex: number): IconCache;
    SetContent(nIndex: number, pControl: IControl): IControl;
    GetContent(nIndex: number): IControl;
    SetItemEnable(nIndex: number, b: boolean): RibbonBackstage;
    GetItemEnable(nIndex: number): boolean;
    GetItemRect(nIndex: number): Rect;

    SetBannerWidth(nWidth: DpiSize): RibbonBackstage;
    GetBannerWidth(): DpiSize;

    SetTextVisible(b: boolean): RibbonBackstage;
    GetTextVisible(): boolean;

    ClearHighlight(): RibbonBackstage;

    OnSelect(
        fn: (sender: RibbonBackstage, nIndex: number) => void
    ): RibbonBackstage;
    OnHighlight(
        fn: (sender: RibbonBackstage, nIndex: number) => void
    ): RibbonBackstage;
}

export class RibbonBackstage extends (AveLib.UiRibbonBackstage as IRibbonBackstage) {
    private m_Content: Array<IControl> = [];

    SetCount(nCount: number) {
        while (this.m_Content.length > nCount) this.m_Content.pop();
        return super.SetCount(nCount);
    }

    SetContent(nIndex: number, pControl: IControl) {
        if (nIndex < this.GetCount()) {
            while (nIndex < this.m_Content.length)
                this.m_Content.push(null as IControl);
            this.m_Content[nIndex] = pControl;
        }
        return super.SetContent(nIndex, pControl);
    }
}

File: Code/Avernakis Nodejs/Avernakis-Nodejs/Private/UiRibbonBackstage.cpp
#include "StdAfx.h"
#include "UiRibbonBackstage.h"
#include "UiRibbonGroup.h"

#define ThisMethod($x) &UiRibbonBackstage::$x
#define AutoAddMethod($x, ...) AddMethod<__VA_ARGS__>( #$x, ThisMethod( $x ) )
#define MakeThisFunc($x) MakeFunc( this, &UiRibbonBackstage::$x )

namespace Nav
{

	namespace
	{
		ObjectRegister<UiRibbonBackstage> c_obj;
	}

	void UiRibbonBackstage::DefineControl()
	{
		AutoAddMethod( SetCount );
		AutoAddMethod( GetCount, WrapObjectGeneric );
		AutoAddMethod( Add );
		AutoAddMethod( Select );
		AutoAddMethod( GetSelection, WrapObjectGeneric );
		AutoAddMethod( GetHighlight, WrapObjectGeneric );
		AutoAddMethod( SetText );
		AutoAddMethod( GetText, WrapObjectGeneric );
		AutoAddMethod( SetIcon );
		AutoAddMethod( GetIcon, WrapObjectGeneric );
		AutoAddMethod( SetContent );
		AutoAddMethod( GetContent, WrapObjectGeneric );
		AutoAddMethod( SetItemEnable );
		AutoAddMethod( GetItemEnable, WrapObjectGeneric );
		AutoAddMethod( GetItemRect, WrapObjectGeneric );
		AutoAddMethod( SetBannerWidth );
		AutoAddMethod( GetBannerWidth, WrapObjectGeneric );
		AutoAddMethod( SetTextVisible );
		AutoAddMethod( GetTextVisible, WrapObjectGeneric );
		AutoAddMethod( ClearHighlight );
		AutoAddMethod( OnSelect, WrapObjectGeneric );
		AutoAddMethod( OnHighlight, WrapObjectGeneric );
	}

	U1 UiRibbonBackstage::Ctor( UiWindow * p, Napi::Value v )
	{
		if ( !__CreateControl( p, v ) )
			return false;

		GetControlTyped().SetTrackingWindowActivity( &p->PublicGetWindow() );
		GetControlTyped().SetIconManager( p->PublicCloneIconManager() );

		return true;
	}

	void UiRibbonBackstage::__OnSelect( Ui::IRibbonBackstage & sender, U32 nIndex )
	{
		m_OnSelect( this, nIndex );
	}

	void UiRibbonBackstage::__OnHighlight( Ui::IRibbonBackstage & sender, S32 nIndex )
	{
		m_OnHighlight( this, nIndex );
	}

	UiRibbonBackstage* UiRibbonBackstage::OnSelect    /**/( Callback_t&& fn ) { m_OnSelect    /**/ = SetEventCallback<Ui::IRibbonBackstage::OnSelect    /**/>( std::move( fn ), MakeThisFunc( __OnSelect    /**/ ) ); return this; }
	UiRibbonBackstage* UiRibbonBackstage::OnHighlight /**/( Callback_t&& fn ) { m_OnHighlight /**/ = SetEventCallback<Ui::IRibbonBackstage::OnHighlight /**/>( std::move( fn ), MakeThisFunc( __OnHighlight /**/ ) ); return this; }

}
